src/db-wrapper.ts

```diff
@@ -0,0 +1,233 @@
+            FOR b IN ${this.blocks}
+            FILTER b._key == ${id}
+            RETURN { 
+                "_key": b._key,
+                "gen_utime": b.gen_utime,
+                "workchain_id": b.workchain_id,
+                "seq_no": b.seq_no,
+                "shard": b.shard,
+                "prev_block_id": b.prev_ref.root_hash,
+                "prev_alt_block_id": b.prev_alt_ref.root_hash,
+                "message_ids": (FOR m_id IN UNION_DISTINCT(
+                        b.out_msg_descr[*].msg_id,
+                        b.out_msg_descr[*].out_msg.msg_id,
+                        b.out_msg_descr[*].reimport.msg_id,
+                        b.in_msg_descr[*].msg_id,
+                        b.in_msg_descr[*].in_msg.msg_id
+                    ) FILTER m_id != null RETURN m_id),
+                "transaction_ids": (FOR t_id IN UNION_DISTINCT(
+                        b.out_msg_descr[*].transaction_id,
+                        b.out_msg_descr[*].reimport.transaction_id
+                    ) FILTER t_id != null RETURN t_id),
+            }
+        `);
+
+        return await query.next();
+    }
+
+    async get_last_masterchain_block_seq_no(): Promise<number | undefined> {
+        const query = await this.arango_db.query(aql`
+            FOR b IN ${this.blocks}
+            FILTER b.workchain_id == -1
+            SORT b.seq_no DESC
+            LIMIT 1
+            RETURN b.seq_no
+        `);
+
+        return await query.next();
+    }
+
+    async init_from(source: IDbWrapper, init_seq_no: number): Promise<void> {
+        const collections = await this.arango_db.collections();
+
+        if (!collections.find(c => c.name == "blocks"))
+            await this.arango_db.createCollection("blocks");
+        
+        if (!collections.find(c => c.name == "messages"))
+            await this.arango_db.createCollection("messages");
+        
+        if (!collections.find(c => c.name == "transactions"))
+            await this.arango_db.createCollection("transactions");
+
+        const masterchain_block_id = await source.get_first_masterchain_block_id_with_seq_no_not_less_than(init_seq_no);
+        if (!masterchain_block_id)
+            throw new Error(`There is no masterchain block with seq_no >= ${init_seq_no} to init from`);
+
+        const masterchain_block = await source.get_masterchain_block_info_by_id(masterchain_block_id);
+        const bmt_ids = {
+            block_ids: [ masterchain_block_id ],
+            message_ids: masterchain_block!.message_ids,
+            transaction_ids: masterchain_block!.transaction_ids,
+        };
+        const bmt = await source.get_BMT_by_ids(bmt_ids);
+        await this.add_BMT(bmt);
+        
+        if (!RangerConfig.test_mode) {
+            await source.remove_BMT_by_ids(bmt_ids);
+        }
+    }
+    
+    async add_BMT(bmt: BMT): Promise<void> {
+        const transaction = await this.arango_db.beginTransaction({
```

Comment:
Are your shure that Arango can accept a huge number of documents per transaction? As a Mitya say we can have abount 10000 TPS and must supports up to 100000 TPS. So you can collect a pool between two master blocks with number of documents about 300000. Just check Arango limitations. May be it is absolutely normal for Arango to work with huge transactions.